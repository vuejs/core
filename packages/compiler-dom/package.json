{
  "name": "@vue/compiler-dom",
  "version": "3.5.13",
  "description": "@vue/compiler-dom",
  "main": "index.js",
  "module": "dist/compiler-dom.esm-bundler.js",
  "types": "dist/compiler-dom.d.ts",
  "unpkg": "dist/compiler-dom.global.js",
  "jsdelivr": "dist/compiler-dom.global.js",
  "files": [
    "index.js",
    "dist"
  ],
  "exports": {
    ".": {
      "types": "./dist/compiler-dom.d.ts",
      "node": {
        "production": "./dist/compiler-dom.cjs.prod.js",
        "development": "./dist/compiler-dom.cjs.js",
        "default": "./index.js"
      },
      "module": "./dist/compiler-dom.esm-bundler.js",
      "import": "./dist/compiler-dom.esm-bundler.js",
      "require": "./index.js"
    },
    "./*": "./*"
  },
  "sideEffects": false,
  "buildOptions": {
    "name": "VueCompilerDOM",
    "compat": true,
    "formats": [
      "esm-bundler",
      "esm-browser",
      "cjs",
      "global"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/vuejs/core.git",
    "directory": "packages/compiler-dom"
  },
  "keywords": [
    "vue"
  ],
  "author": "Evan You",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/vuejs/core/issues"
  },
  "homepage": "https://github.com/vuejs/core/tree/main/packages/compiler-dom#readme",
  "dependencies": {
    "@vue/shared": "workspace:*",
    "@vue/compiler-core": "workspace:*"
  }
}
