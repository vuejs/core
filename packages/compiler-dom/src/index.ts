import {
  type CodegenResult,
  type CompilerOptions,
  type DirectiveTransform,
  type NodeTransform,
  type ParserOptions,
  type RootNode,
  baseCompile,
  baseParse,
  noopDirectiveTransform,
} from '@vue/compiler-core'
import { parserOptions } from './parserOptions'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformModel } from './transforms/vModel'
import { transformOn } from './transforms/vOn'
import { transformShow } from './transforms/vShow'
import { transformTransition } from './transforms/Transition'
import { stringifyStatic } from './transforms/stringifyStatic'
import { ignoreSideEffectTags } from './transforms/ignoreSideEffectTags'
import { extend } from '@vue/shared'

export { parserOptions }

export const DOMNodeTransforms: NodeTransform[] = [
  transformStyle,
  ...(__DEV__ ? [transformTransition] : []),
]

export const DOMDirectiveTransforms: Record<string, DirectiveTransform> = {
  cloak: noopDirectiveTransform,
  html: transformVHtml,
  text: transformVText,
  model: transformModel, // override compiler-core
  on: transformOn, // override compiler-core
  show: transformShow,
}

export function compile(
  src: string | RootNode,
  options: CompilerOptions = {},
): CodegenResult {
  return baseCompile(
    src,
    extend({}, parserOptions, options, {
      nodeTransforms: [
        // ignore <script> and <tag>
        // this is not put inside DOMNodeTransforms because that list is used
        // by compiler-ssr to generate vnode fallback branches
        ignoreSideEffectTags,
        ...DOMNodeTransforms,
        ...(options.nodeTransforms || []),
      ],
      directiveTransforms: extend(
        {},
        DOMDirectiveTransforms,
        options.directiveTransforms || {},
      ),
      transformHoist: __BROWSER__ ? null : stringifyStatic,
    }),
  )
}

export function parse(template: string, options: ParserOptions = {}): RootNode {
  return baseParse(template, extend({}, parserOptions, options))
}

export * from './runtimeHelpers'
export { transformStyle } from './transforms/transformStyle'
export {
  createDOMCompilerError,
  DOMErrorCodes,
  DOMErrorMessages,
} from './errors'
export * from '@vue/compiler-core'
