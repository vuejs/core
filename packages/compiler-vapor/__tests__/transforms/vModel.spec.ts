import { makeCompile } from './_utils'
import {
  IRNodeTypes,
  transformChildren,
  transformElement,
  transformVModel,
} from '../../src'
import { BindingTypes, DOMErrorCodes } from '@vue/compiler-dom'

const compileWithVModel = makeCompile({
  nodeTransforms: [transformElement, transformChildren],
  directiveTransforms: {
    model: transformVModel,
  },
})

describe('compiler: vModel transform', () => {
  test('should support simple expression', () => {
    const { code, helpers } = compileWithVModel('<input v-model="model" />')
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyTextModel')
  })

  test('should support input (text)', () => {
    const { code, helpers } = compileWithVModel(
      '<input type="text" v-model="model" />',
    )
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyTextModel')
  })

  test('should support input (radio)', () => {
    const { code, helpers } = compileWithVModel(
      '<input type="radio" v-model="model" />',
    )
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyRadioModel')
  })

  test('should support input (checkbox)', () => {
    const { code, helpers } = compileWithVModel(
      '<input type="checkbox" v-model="model" />',
    )
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyCheckboxModel')
  })

  test('should support select', () => {
    const { code, helpers } = compileWithVModel('<select v-model="model" />')
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applySelectModel')
  })

  test('should support textarea', () => {
    const { code, helpers } = compileWithVModel('<textarea v-model="model" />')
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyTextModel')
  })

  test('should support input (dynamic type)', () => {
    const { code, helpers } = compileWithVModel(
      '<input :type="foo" v-model="model" />',
    )
    expect(code).toMatchSnapshot()
    expect(helpers).toContain('applyDynamicModel')
  })

  test('should support w/ dynamic v-bind', () => {
    const root1 = compileWithVModel('<input v-bind="obj" v-model="model" />')
    expect(root1.code).toMatchSnapshot()
    expect(root1.helpers).toContain('applyDynamicModel')

    const root2 = compileWithVModel(
      '<input v-bind:[key]="val" v-model="model" />',
    )
    expect(root2.code).toMatchSnapshot()
    expect(root2.helpers).toContain('applyDynamicModel')
  })

  describe('errors', () => {
    test('invalid element', () => {
      const onError = vi.fn()
      compileWithVModel('<span v-model="model" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ON_INVALID_ELEMENT,
        }),
      )
    })

    test('plain elements with argument', () => {
      const onError = vi.fn()
      compileWithVModel('<input v-model:value="model" />', { onError })

      expect(onError).toHaveBeenCalledTimes(1)
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ARG_ON_ELEMENT,
        }),
      )
    })

    // TODO: component
    test.todo('should allow usage on custom element', () => {
      const onError = vi.fn()
      const root = compileWithVModel('<my-input v-model="model" />', {
        onError,
        isCustomElement: tag => tag.startsWith('my-'),
      })
      expect(root.helpers).toContain('applyTextModel')
      expect(onError).not.toHaveBeenCalled()
    })

    test('should raise error if used file input element', () => {
      const onError = vi.fn()
      compileWithVModel(`<input type="file" v-model="test"/>`, {
        onError,
      })
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_ON_FILE_INPUT_ELEMENT,
        }),
      )
    })

    test('should error on dynamic value binding alongside v-model', () => {
      const onError = vi.fn()
      compileWithVModel(`<input v-model="test" :value="test" />`, {
        onError,
      })
      expect(onError).toHaveBeenCalledWith(
        expect.objectContaining({
          code: DOMErrorCodes.X_V_MODEL_UNNECESSARY_VALUE,
        }),
      )
    })

    // #3596
    test('should NOT error on static value binding alongside v-model', () => {
      const onError = vi.fn()
      compileWithVModel(`<input v-model="test" value="test" />`, {
        onError,
      })
      expect(onError).not.toHaveBeenCalled()
    })
  })

  describe('modifiers', () => {
    test('.number', () => {
      const { code } = compileWithVModel('<input v-model.number="model" />')

      expect(code).toMatchSnapshot()
    })

    test('.trim', () => {
      const { code } = compileWithVModel('<input v-model.trim="model" />')

      expect(code).toMatchSnapshot()
    })

    test('.lazy', () => {
      const { code } = compileWithVModel('<input v-model.lazy="model" />')

      expect(code).toMatchSnapshot()
    })
  })

  test('should support member expression', () => {
    const { code } = compileWithVModel(
      '<input v-model="setupRef.child" /><input v-model="setupLet.child" /><input v-model="setupMaybeRef.child" />',
      {
        bindingMetadata: {
          setupRef: BindingTypes.SETUP_REF,
          setupLet: BindingTypes.SETUP_LET,
          setupMaybeRef: BindingTypes.SETUP_MAYBE_REF,
        },
      },
    )

    expect(code).toMatchSnapshot()
  })

  test('should support member expression w/ inline', () => {
    const { code } = compileWithVModel(
      '<input v-model="setupRef.child" /><input v-model="setupLet.child" /><input v-model="setupMaybeRef.child" />',
      {
        bindingMetadata: {
          setupRef: BindingTypes.SETUP_REF,
          setupLet: BindingTypes.SETUP_LET,
          setupMaybeRef: BindingTypes.SETUP_MAYBE_REF,
        },
        inline: true,
      },
    )

    expect(code).toMatchSnapshot()
  })

  describe('component', () => {
    test('v-model for component should work', () => {
      const { code, ir } = compileWithVModel('<Comp v-model="foo" />')
      expect(code).toMatchSnapshot()
      expect(code).contains(`modelValue: () => (_ctx.foo),`)
      expect(code).contains(
        `"onUpdate:modelValue": () => _value => (_ctx.foo = _value)`,
      )
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          [
            {
              key: { content: 'modelValue', isStatic: true },
              model: true,
              modelModifiers: [],
              values: [{ content: 'foo', isStatic: false }],
            },
          ],
        ],
      })
    })

    test('v-model with arguments for component should work', () => {
      const { code, ir } = compileWithVModel('<Comp v-model:bar="foo" />')
      expect(code).toMatchSnapshot()
      expect(code).contains(`bar: () => (_ctx.foo),`)
      expect(code).contains(
        `"onUpdate:bar": () => _value => (_ctx.foo = _value)`,
      )
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          [
            {
              key: { content: 'bar', isStatic: true },
              model: true,
              modelModifiers: [],
              values: [{ content: 'foo', isStatic: false }],
            },
          ],
        ],
      })
    })

    test('v-model with dynamic arguments for component should work', () => {
      const { code, ir } = compileWithVModel('<Comp v-model:[arg]="foo" />')
      expect(code).toMatchSnapshot()
      expect(code).contains(
        `[_ctx.arg]: _ctx.foo,
    ["onUpdate:" + _ctx.arg]: () => _value => (_ctx.foo = _value)`,
      )
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          {
            key: { content: 'arg', isStatic: false },
            values: [{ content: 'foo', isStatic: false }],
            model: true,
            modelModifiers: [],
          },
        ],
      })
    })

    test('v-model for component should generate modelModifiers', () => {
      const { code, ir } = compileWithVModel(
        '<Comp v-model.trim.bar-baz="foo" />',
      )
      expect(code).toMatchSnapshot()
      expect(code).contain(
        `modelModifiers: () => ({ trim: true, "bar-baz": true })`,
      )
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          [
            {
              key: { content: 'modelValue', isStatic: true },
              values: [{ content: 'foo', isStatic: false }],
              model: true,
              modelModifiers: ['trim', 'bar-baz'],
            },
          ],
        ],
      })
    })

    test('v-model with arguments for component should generate modelModifiers', () => {
      const { code, ir } = compileWithVModel(
        '<Comp v-model:foo.trim="foo" v-model:bar.number="bar" />',
      )
      expect(code).toMatchSnapshot()
      expect(code).contain(`fooModifiers: () => ({ trim: true })`)
      expect(code).contain(`barModifiers: () => ({ number: true })`)
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          [
            {
              key: { content: 'foo', isStatic: true },
              values: [{ content: 'foo', isStatic: false }],
              model: true,
              modelModifiers: ['trim'],
            },
            {
              key: { content: 'bar', isStatic: true },
              values: [{ content: 'bar', isStatic: false }],
              model: true,
              modelModifiers: ['number'],
            },
          ],
        ],
      })
    })

    test('v-model with dynamic arguments for component should generate modelModifiers ', () => {
      const { code, ir } = compileWithVModel(
        '<Comp v-model:[foo].trim="foo" v-model:[bar].number="bar" />',
      )
      expect(code).toMatchSnapshot()
      expect(code).contain(`[_ctx.foo + "Modifiers"]: () => ({ trim: true })`)
      expect(code).contain(`[_ctx.bar + "Modifiers"]: () => ({ number: true })`)
      expect(ir.block.dynamic.children[0].operation).toMatchObject({
        type: IRNodeTypes.CREATE_COMPONENT_NODE,
        tag: 'Comp',
        props: [
          {
            key: { content: 'foo', isStatic: false },
            values: [{ content: 'foo', isStatic: false }],
            model: true,
            modelModifiers: ['trim'],
          },
          {
            key: { content: 'bar', isStatic: false },
            values: [{ content: 'bar', isStatic: false }],
            model: true,
            modelModifiers: ['number'],
          },
        ],
      })
    })
  })
})
