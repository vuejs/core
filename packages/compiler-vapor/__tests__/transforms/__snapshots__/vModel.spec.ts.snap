// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: vModel transform > component > v-model for component should generate modelModifiers 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, { modelValue: () => (_ctx.foo),
  "onUpdate:modelValue": () => _value => (_ctx.foo = _value),
  modelModifiers: () => ({ trim: true, "bar-baz": true }) }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > component > v-model for component should work 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, { modelValue: () => (_ctx.foo),
  "onUpdate:modelValue": () => _value => (_ctx.foo = _value) }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > component > v-model with arguments for component should generate modelModifiers 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, {
    foo: () => (_ctx.foo),
    "onUpdate:foo": () => _value => (_ctx.foo = _value),
    fooModifiers: () => ({ trim: true }), 
    bar: () => (_ctx.bar),
    "onUpdate:bar": () => _value => (_ctx.bar = _value),
    barModifiers: () => ({ number: true })
  }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > component > v-model with arguments for component should work 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, { bar: () => (_ctx.foo),
  "onUpdate:bar": () => _value => (_ctx.foo = _value) }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > component > v-model with dynamic arguments for component should generate modelModifiers  1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, { $: [
    () => ({ [_ctx.foo]: _ctx.foo,
    ["onUpdate:" + _ctx.foo]: () => _value => (_ctx.foo = _value),
    [_ctx.foo + "Modifiers"]: () => ({ trim: true }) }), 
    () => ({ [_ctx.bar]: _ctx.bar,
    ["onUpdate:" + _ctx.bar]: () => _value => (_ctx.bar = _value),
    [_ctx.bar + "Modifiers"]: () => ({ number: true }) })
  ] }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > component > v-model with dynamic arguments for component should work 1`] = `
"import { resolveComponent as _resolveComponent, createComponentWithFallback as _createComponentWithFallback } from 'vue';

export function render(_ctx) {
  const _component_Comp = _resolveComponent("Comp")
  const n0 = _createComponentWithFallback(_component_Comp, { $: [
    () => ({ [_ctx.arg]: _ctx.foo,
    ["onUpdate:" + _ctx.arg]: () => _value => (_ctx.foo = _value) })
  ] }, null, true)
  return n0
}"
`;

exports[`compiler: vModel transform > modifiers > .lazy 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value), { lazy: true })
  return n0
}"
`;

exports[`compiler: vModel transform > modifiers > .number 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value), { number: true })
  return n0
}"
`;

exports[`compiler: vModel transform > modifiers > .trim 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value), { trim: true })
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (checkbox) 1`] = `
"import { applyCheckboxModel as _applyCheckboxModel, template as _template } from 'vue';
const t0 = _template("<input type=\\"checkbox\\">", true)

export function render(_ctx) {
  const n0 = t0()
  _applyCheckboxModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (dynamic type) 1`] = `
"import { applyDynamicModel as _applyDynamicModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyDynamicModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (radio) 1`] = `
"import { applyRadioModel as _applyRadioModel, template as _template } from 'vue';
const t0 = _template("<input type=\\"radio\\">", true)

export function render(_ctx) {
  const n0 = t0()
  _applyRadioModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support input (text) 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input type=\\"text\\">", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support member expression 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input>")

export function render(_ctx, $props, $emit, $attrs, $slots) {
  const n0 = t0()
  const n1 = t0()
  const n2 = t0()
  _applyTextModel(n0, () => (_ctx.setupRef.child), _value => (_ctx.setupRef.child = _value))
  _applyTextModel(n1, () => (_ctx.setupLet.child), _value => (_ctx.setupLet.child = _value))
  _applyTextModel(n2, () => (_ctx.setupMaybeRef.child), _value => (_ctx.setupMaybeRef.child = _value))
  return [n0, n1, n2]
}"
`;

exports[`compiler: vModel transform > should support member expression w/ inline 1`] = `
"
  const n0 = t0()
  const n1 = t0()
  const n2 = t0()
  _applyTextModel(n0, () => (setupRef.value.child), _value => (setupRef.value.child = _value))
  _applyTextModel(n1, () => (_unref(setupLet).child), _value => (_unref(setupLet).child = _value))
  _applyTextModel(n2, () => (_unref(setupMaybeRef).child), _value => (_unref(setupMaybeRef).child = _value))
  return [n0, n1, n2]
"
`;

exports[`compiler: vModel transform > should support select 1`] = `
"import { applySelectModel as _applySelectModel, template as _template } from 'vue';
const t0 = _template("<select></select>", true)

export function render(_ctx) {
  const n0 = t0()
  _applySelectModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support simple expression 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support textarea 1`] = `
"import { applyTextModel as _applyTextModel, template as _template } from 'vue';
const t0 = _template("<textarea></textarea>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyTextModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;

exports[`compiler: vModel transform > should support w/ dynamic v-bind 1`] = `
"import { applyDynamicModel as _applyDynamicModel, setDynamicProps as _setDynamicProps, renderEffect as _renderEffect, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyDynamicModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  _renderEffect(() => _setDynamicProps(n0, [_ctx.obj], true))
  return n0
}"
`;

exports[`compiler: vModel transform > should support w/ dynamic v-bind 2`] = `
"import { applyDynamicModel as _applyDynamicModel, template as _template } from 'vue';
const t0 = _template("<input>", true)

export function render(_ctx) {
  const n0 = t0()
  _applyDynamicModel(n0, () => (_ctx.model), _value => (_ctx.model = _value))
  return n0
}"
`;
